continuousscanning/cooldownqueue.go

```diff
@@ -0,0 +1,81 @@
+package continuousscanning
+
+import (
+	"context"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru/v2/expirable"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/watch"
+
+	"github.com/kubescape/go-logger"
+	"github.com/kubescape/go-logger/helpers"
+)
+
+const (
+	// Default size for the cooldown queue
+	DefaultQueueSize = 512
+	// Default TTL for events put in the queue
+	DefaultTTL = 5 * time.Second
+)
+
+// cooldownQueue is a queue that lets clients put events into it with a cooldown
+//
+// When a client puts an event into a queue, it forwards the event to its
+// output channel and starts a cooldown for this event. If a client attempts to
+// put the same event into the queue while the cooldown is running, the queue
+// will silently drop the event. When the cooldown resets and a client puts the
+// same event into the queue, it will be forwarded to the output channel
+type cooldownQueue struct {
+	seenEvents *lru.LRU[string, bool]
+	// inner channel for producing events
+	innerChan chan watch.Event
+	// public channel for reading events
+	ResultChan <-chan watch.Event
+}
+
+// NewCooldownQueue returns a new Cooldown Queue
+func NewCooldownQueue(size int, cooldown time.Duration) *cooldownQueue {
+	lru := lru.NewLRU[string, bool](size, nil, cooldown)
+	c := make(chan watch.Event)
+
+	return &cooldownQueue{
+		seenEvents: lru,
+		innerChan:  c,
+		ResultChan: c,
+	}
+
+}
+
+func makeEventKey(e watch.Event) string {
+	object, ok := e.Object.(*unstructured.Unstructured)
+	if !ok {
+		return ""
+	}
+
+	eventKey := string(e.Type) + "-" + string(object.GetUID())
```

Comment:
I think we can remove the type
```suggestion
	eventKey := string(object.GetUID())
```

Reply:
No, we cannot.

If we create a pod and then delete it, the cooldown queue will discard the DELETED event. This will prevent the handler that removes CRDs from triggering a Kubescape deleted-resource scan.